pkg/scheduling/volumelimits.go

```diff
@@ -0,0 +1,94 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// VolumeLimits tracks volume limits on a per node basis.  The number of volumes that can be mounted varies by instance
+// type. We need to be aware and track the mounted volume usage to inform our awareness of which pods can schedule to
+// which nodes.
+type VolumeLimits struct {
+	ctx        context.Context
+	volumes    sets.String
+	podVolumes map[types.NamespacedName]sets.String
+}
+
+func NewVolumeLimits(ctx context.Context) *VolumeLimits {
+	return &VolumeLimits{
+		ctx:        ctx,
+		volumes:    sets.NewString(),
+		podVolumes: map[types.NamespacedName]sets.String{},
+	}
+}
+
+func (v *VolumeLimits) Add(pod *v1.Pod) {
+	podVolumes, err := v.validate(pod)
+	if err != nil {
+		logging.FromContext(v.ctx).Errorf("inconsistent state registering volume li mit, %s", err)
+	}
+	v.podVolumes[client.ObjectKeyFromObject(pod)] = podVolumes
+	v.volumes = v.volumes.Union(podVolumes)
+}
+
+func (v *VolumeLimits) Validate(pod *v1.Pod) (int, error) {
+	podVolumes, err := v.validate(pod)
+	result := v.volumes.Union(podVolumes)
+	return len(result), err
+}
+
+func (v *VolumeLimits) validate(pod *v1.Pod) (sets.String, error) {
```

Comment:
Looks like this can never return an error?

Reply:
It did previously, I was validating the PVC and volume, but decided not to check the volume.  Removing the error.